File: src/medium/NumIslands.java
package medium;

/**
 * https://leetcode.com/problems/number-of-islands/solution/
 */
public class NumIslands {
    public int NumIslands(char[][] grid) {
        int islandCount = 0;
        int rowLen = grid.length;
        if(rowLen < 1) {
            return 0;
        }
        int colLen = grid[0].length;
        for (int i=0; i<rowLen; i++){
            for (int j=0; j<colLen; j++){
                if(grid[i][j] == '1') {
                    ++islandCount;
                    dfs(grid, i, j);
                }
            }
        }
        return islandCount;
    }

    private void dfs(char[][] grid, int i, int j){
        // mark visited point to 0
        grid[i][j] = 0;
        //check left
        if (j-1>= 0 && grid[i][j-1] == '1'){
            dfs(grid, i, j-1);
        }
        //check right
        if (j+1<grid[0].length && grid[i][j+1] == '1'){
            dfs(grid, i, j+1);
        }
        //check top
        if (i-1>=0 && grid[i-1][j] == '1'){
            dfs(grid, i-1, j);
        }
        //check bottom
        if (i+1< grid.length && grid[i+1][j] == '1'){
            dfs(grid, i+1, j);
        }

    }
}

File: csharp/LeetCode/Medium/NumIslands.cs
namespace LeetCode.Medium
{
    public partial class Solution
    {

        public int NumIslands(char[][] grid)
        {
            if (grid == null || grid.Length == 0 || grid[0].Length == 0)
            {
                return 0;
            }

            var m = grid.Length;
            var n = grid[0].Length;
            var i = 0;
            var visited = new bool[m, n];
            var count = 0;
            while (i < m)
            {
                var j = 0;
                while (j < n)
                {
                    if (!visited[i, j])
                    {
                        this.Visit(i, j, grid, visited);
                        if (grid[i][j] == '1')
                        {
                            count++;
                        }
                    }
                    j++;
                }
                i++;
            }
            return count;
        }

        private void Visit(int i, int j, char[][] grid, bool[,] visited)
        {
            if (i >= 0 && j >= 0 && i < grid.Length && j < grid[0].Length && !visited[i, j])
            {
                visited[i, j] = true;
                if (grid[i][j] == '1')
                {
                    // go left
                    this.Visit(i, j - 1, grid, visited);
                    // go right
                    this.Visit(i, j + 1, grid, visited);
                    // go up
                    this.Visit(i - 1, j, grid, visited);
                    // go bottom
                    this.Visit(i + 1, j, grid, visited);
                }
                // not a land
            }
        }
    }
}
